Add support for string and time indexed event stats

MatMeredith opened the pull request:
Tests https://github.com/Metaswitch/cpp-common/pull/658

MatMeredith committed:
Improve testing of EventAccumulator tables

MatMeredith committed:
Add time and string based event table

MatMeredith committed:
Add tests for time and string based event table

MatMeredith committed:
Don't need to use different OIDs

johadalin commented on src/test_snmp.cpp:
```diff
@@ -12,6 +12,8 @@
 
 #include "test_snmp.h"
+#include <string>
+
```
blank line

johadalin commented on src/test_snmp.cpp:
```diff
@@ -86,6 +88,190 @@ TEST_F(SNMPTest, LatencyCalculations)
+
+  // Create a table.  Use a different OID base because otherwise the indices
+  // clash with later tests.
+  SNMP::EventAccumulatorByScopeTable* tbl = SNMP::EventAccumulatorByScopeTable::create("latency", test_oid);
```
You say 'use a different OID base', but then go on to use the same `test_oid` as elsewhere. Am i missing something here?

johadalin replied:
Just seen the commit comment saying no need to use different OIDs. Can you update the comments to reflect this?

johadalin commented on src/test_snmp.cpp:
```diff
@@ -86,6 +88,190 @@ TEST_F(SNMPTest, LatencyCalculations)
+  tbl->accumulate(100);
+
+  // Check the counts
+  ASSERT_EQ(0, snmp_get(".1.2.2.1.7.1.4.110.111.100.101"));
```
Can you add comments for the other two OIDs here as well? and those below. Seems really useful to give them human readable values

johadalin commented on src/test_snmp.cpp:
```diff
@@ -86,6 +88,190 @@ TEST_F(SNMPTest, LatencyCalculations)
+  ASSERT_EQ(300, snmp_get(".1.2.2.1.6.1.4.110.111.100.101"));
+  // Count should be 2
+  ASSERT_EQ(2, snmp_get(".1.2.2.1.7.1.4.110.111.100.101"));
+
```
Why no test for variance here?

johadalin commented on src/test_snmp.cpp:
```diff
@@ -86,6 +88,190 @@ TEST_F(SNMPTest, LatencyCalculations)
+  delete tbl;
+}
+
+enum EventStats
```
Are we able to use these ENUM values in the test above to make it more readable?

johadalin commented on src/test_snmp.cpp:
```diff
@@ -86,6 +88,190 @@ TEST_F(SNMPTest, LatencyCalculations)
+{
+  cwtest_completely_control_time(true);
+
+  // Create a table.  Use a different OID base because otherwise the indices?// clash with later tests.
```
Feel like this is a typo here. Also, as above, you're using the same `test_oid`.

johadalin commented on src/test_snmp.cpp:
```diff
@@ -86,6 +88,190 @@ TEST_F(SNMPTest, LatencyCalculations)
+  std::string string2_oid = "16.115.101.114.118.101.114.50.46.100.99.49.46.122.111.110.101";
+  tbl->accumulate(string1, 100);
+
+  //snmp_walk_debug(".1.2.2");
```
If this is just for debugging what's going on, it should be removed

johadalin replied:
Also, is this whole test function just for debugging? If so, shouldn't really be in the final code, right?

MatMeredith replied:
The function is just for debugging, but it's not going to be the last time that I (or someone else working on this wants it).   This isn't production code, so is there a strong argument for removing it?

johadalin commented on src/test_snmp.cpp:
```diff
@@ -86,6 +88,190 @@ TEST_F(SNMPTest, LatencyCalculations)
+
+  // Check that there are the right number of entries in the table (3 time
+  // periods * 1 string index * 5 metrics)
+  std::vector<std::string> entries = snmp_walk(".1.2.2");
```
is .1.2.2 the test_oid value? can we use `test_oid` here instead?

johadalin reviewed (changes requested):
The tests seem decent, but there are a number of places where the code just needs a clean up. Can you take another look over it to make sure comments are correct, and you've cleaned up the debug steps you put in.

MatMeredith commented:
@johadalin Thanks -- sorry -- this was a bit of a rushed check in.   I've addressed the most blatant issues that you've pointed out but there is a lot more that could be done in terms of making existing tests better.   The main thing I'd like to do is to generalise the time_string_event_oid function and use this throughout the other tests to make the OIDs we're testing much more readable as I have done with the new test.   

I'll come back to this if I get time tomorrow, otherwise I'm happy that the new test I've added raises the quality bar compared to what was already there, and that I've also made stuff better (e.g. adding the LatencyScopeCalculations test that was previously missing and making it more thorough than the LatencyCalculations test that it is copied from)!

Is that OK?

MatMeredith committed:
Code review markups

johadalin commented:
Sounds like a reasonable plan. All looks good as is to me. Happy to leave generalising the time_string_event_oid function as a future enhancement, possibly tracked in an issue or somehting somewhere. Otherwise this looks good to merge now :+1:
